src/bin/pgcopydb/ld_replay.c

```diff
@@ -249,3 +250,136 @@ stream_replay_line(void *ctx, const char *line, bool *stop)
 
 	return true;
 }
+
+
+/* streamReplayLineIntoBuffer is a callback function for the ReadFromStreamContext
+ * and read_from_stream infrastructure. It's called on each line read from a
+ * stream such as a unix pipe and buffers the line into a file until we encounter
+ * a COMMIT. This ensures that we can replay the whole transaction in one go
+ * and avoids deadlock when the BEGIN messages arrives before the existence of
+ * transation metadata file which is created by the transform process.
+ */
+static bool
+streamReplayLineIntoBuffer(void *ctx, const char *line, bool *stop)
+{
+	ReplayStreamCtx *replayCtx = (ReplayStreamCtx *) ctx;
+	StreamApplyContext *context = &(replayCtx->applyContext);
+
+	if (strstr(line, OUTPUT_BEGIN) == line)
+	{
+		/* If there is already a replay buffer open, it indicates a partial
+		 * transaction. Close it and start a new one.
+		 */
+		if (replayCtx->txnReplayBuffer != NULL)
+		{
+			fclose(replayCtx->txnReplayBuffer);
+		}
+
+		LogicalMessageMetadata metadata = { 0 };
+		if (!parseSQLAction((char *) line, &metadata))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		/* Create a txid as a symlink to the buffer file */
+		char txidFile[MAXPGPATH];
+		sformat(txidFile, MAXPGPATH, "%s/%d.sql", context->paths.dir, metadata.xid);
```

Comment:
Again, let's expose the current name as `replayCtx->currentXIDfilename` or something.